### ISSUE TITLE
trailing comma in __geo_interface__

### ISSUE THREAD

**Issue by simonm3** [geopandas/geopandas#1551]
__geo_interface__ and the synonym _to_geo() add a trailing comma at the end of the coordinates. This is not valid json and can cause failure e.g. it does not work with mapboxgl. Also it is unclear why this is an __ function. Would be clearer if it were called to_geo()

The workaround is to do json.loads(df.to_json()). This does not add a trailing comma though I am not sure if the results are always identical.

**Comment by martinfleis** [comment]
Thanks for the report! The trailing comma is returned by shapely in `mapping(geom)`, so the issue should be probably reported and fixed there. We could do some cleaning on our side, but I'd prefer to fix the cause of it.

 `_to_geo()` is a private function, not a part of a public API. That is why it starts with an underscore. There is a discussion on making it public in #1320.

The result from `__geo_interface__` is the same as from `to_json()` as both are generated using `to_geo()`, the only difference is that the result gets roundtripped through `json.dumps()` in `to_json()` and `json.loads()`, which probably does some sanitization (i.e. remove the comma).

**Comment by simonm3** [comment]
See https://github.com/Toblerity/Shapely/issues/953

**Comment by martinfleis** [comment]
In the light of Toblerity/Shapely#953, we may actually put some effort into returning valid JSON in GeoPandas.

**Comment by jorisvandenbossche** [comment]
About which trailing comma are we exactly speaking?

Eg using this example:

```
In [12]: import geopandas 

In [13]: from shapely.geometry import Point 

In [14]: gdf = geopandas.GeoDataFrame({'a': [0, 1], 'geometry': [Point(0, 0), Point(1, 1)]}) 

In [15]: gdf._to_geo() 
Out[15]: 
{'type': 'FeatureCollection',
 'features': [{'id': '0',
   'type': 'Feature',
   'properties': {'a': 0},
   'geometry': {'type': 'Point', 'coordinates': (0.0, 0.0)}},
  {'id': '1',
   'type': 'Feature',
   'properties': {'a': 1},
   'geometry': {'type': 'Point', 'coordinates': (1.0, 1.0)}}]}
```

Here, I don't see a trailing comma at the end of the coordinates.

**Comment by jorisvandenbossche** [comment]
Ah, using a polygon example instead of point, I see the trailing comma:

```
In [22]: from shapely.geometry import box 

In [23]: gdf = geopandas.GeoDataFrame({'a': [0], 'geometry': [box(0, 0, 1, 1)]}) 

In [24]: gdf._to_geo() 
Out[24]: 
{'type': 'FeatureCollection',
 'features': [{'id': '0',
   'type': 'Feature',
   'properties': {'a': 0},
   'geometry': {'type': 'Polygon',
    'coordinates': (((1.0, 0.0),
      (1.0, 1.0),
      (0.0, 1.0),
      (0.0, 0.0),
      (1.0, 0.0)),)}}]}
```

But this the python repr of a tuple (there is a tuple of length 1, and then it uses a trailing comma to indicate it is actually a tuple, like `(1,)`). See:

```
In [30]: mapping(box(0, 0, 1, 1))['coordinates'] 
Out[30]: (((1.0, 0.0), (1.0, 1.0), (0.0, 1.0), (0.0, 0.0), (1.0, 0.0)),)
```

If we have multiple rings in that tuple (eg if there is a hole), you no longer have this trailing comma:

```
In [31]: mapping(box(0, 0, 3, 3).difference(box(1, 1, 2, 2)))['coordinates'] 
Out[31]: 
(((3.0, 0.0), (0.0, 0.0), (0.0, 3.0), (3.0, 3.0), (3.0, 0.0)),
 ((2.0, 1.0), (2.0, 2.0), (1.0, 2.0), (1.0, 1.0), (2.0, 1.0)))
```

Further, if you actually dump the python  dict with tuple with trailing comma to json, it already correctly formats this without any trailing comma:

```
In [33]: import json                                                                                                                                                                                               

In [34]: json.dumps(gdf._to_geo())                                                                                                                                                                                 
Out[34]: '{"type": "FeatureCollection", "features": [{"id": "0", "type": "Feature", "properties": {"a": 0}, "geometry": {"type": "Polygon", "coordinates": [[[1.0, 0.0], [1.0, 1.0], [0.0, 1.0], [0.0, 0.0], [1.0, 0.0]]]}}]}'
```

**Comment by simonm3** [comment]
Yes it is easy enough to do that. However it would be logical for the
geodataframe and geoseries to have a "geojson" methods that returns valid
geojson with no trailing comma and with the CRS.

Something like this:

def geojson(x):
    """ return geojson from geodataframe or geoseries
    includes geometry, crs, properties (from other columns)
    """
    f = BytesIO()
    x.to_file(f, driver="GeoJSON")
    return json.loads(f.getvalue())

**Comment by jorisvandenbossche** [comment]
Can you explain your use case a bit? 
As again, we currenlty don't return "invalid" geojson, we return a python dict that gets converted to json correctly with `json.dump(s)`. In which case is the "trailing comma" (so actually the fact that it is a tuple and not a list) a problem?

**Comment by simonm3** [comment]
Mapbox and plotly mapbox require geojson. They fail with the trailing comma. I suspect there would be problems if you were using a crs other than 4326 as the crs is missing from the geo function too.

Also if you have a function called geo one might expect it to return geojson not a dict.

**Comment by jorisvandenbossche** [comment]
Can you give a concrete code example how the output of geopandas is passed to mapbox of plotly? (not trying to be pedantic, just trying to understand the use case, to be able to help / to see what would be useful to change in geopandas)

**Comment by simonm3** [comment]
I cannot find the plotly example now where it failed - it seems to work in many cases even with the extra comma. It still fails in mapbox using __geo_data__ but works fine using fiona as per the function I used above. I don't have simple example as it is in python and javascript. I would just drop that in because it works and the obvious behaviour should be to return valid geojson rather than a geo dict or invalid json without the crs.

**Comment by simonm3** [comment]
OK here is one. The HTML at the top works. Replacing the map.addSource
section with the one from __geo_interface__ below fails.
There are several differences but the correct geojson that works:

   - is valid json with no trailing comma
   - has a crs
   - uses lists rather than tuples
   - may be other differences

But in the end may as well use what works which is the fiona version.

<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8" />
    <title>Display a map</title>
    <meta name="viewport"
content="initial-scale=1,maximum-scale=1,user-scalable=no" />
    <script src="https://api.mapbox.com/mapbox-gl-js/v1.11.1/mapbox-gl.js
"></script>
    <link href="https://api.mapbox.com/mapbox-gl-js/v1.11.1/mapbox-gl.css"
rel="stylesheet" />
<style>
    #map {
    position: absolute;
    top: 30px;
    bottom: 0;
    width: 100%;
}
</style>
</head>

<body>
    <div id="map"></div>


    <script>

        mapboxgl.accessToken =
'pk.eyJ1Ijoic2ltb25tMyIsImEiOiJja2Q1eDd0bzIwMjFuMnlzYXJ2cTV3MTJsIn0.RI-Oc6sEd0JL3iECMXU8kQ';
        var map = new mapboxgl.Map(
            {'container': 'map', 'style':
'mapbox://styles/mapbox/streets-v11', 'center': [-1.5, 53], 'zoom': 5}
        );
        map.on('load', function () {
            // sources

           map.addSource('ukdistricts', {'data': {'type':
'FeatureCollection', 'crs': {'type': 'name', 'properties': {'name':
'urn:ogc:def:crs:OGC:1.3:CRS84'}}, 'features': [{'type': 'Feature',
'properties': {'authority': 'Hartlepool', 'names': 0.0, 'registr':
'non-user'}, 'geometry': {'type': 'MultiPolygon', 'coordinates':
[[[[-1.268441173945158, 54.726106057183536], [-1.177618431962607,
54.69917558358339], [-1.199082154185432, 54.68053931094894],
[-1.160815262157439, 54.63470116747878], [-1.256702670448064,
54.623338735223086], [-1.298999444034391, 54.6279702152551],
[-1.316349741584692, 54.64522360530093], [-1.335046640099353,
54.63131255558137], [-1.382681311227283, 54.64438440310203],
[-1.341239879139743, 54.65020086862752], [-1.34451182787862,
54.69111059347379], [-1.268441173945158, 54.726106057183536]]],
[[[-1.24214073682385, 54.72239561694049], [-1.242505104062029,
54.722420353944365], [-1.241920522611265, 54.7227132713354],
[-1.241904448892593, 54.722497548992735], [-1.24214073682385,
54.72239561694049]]], [[[-1.224653382900811, 54.62611617282496],
[-1.224687817795711, 54.62609842464633], [-1.22449128717292,
54.62617173704485], [-1.224653382900811, 54.62611617282496]]]]}}]}, 'type':
'geojson'});

  // layers

            map.addLayer({'type': 'fill', 'source': 'ukdistricts',
'property': 'registr'});
        });
    </script>

</body>

</html>

#######################################
map.addSource('ukdistricts', {'data': {'type': 'FeatureCollection',
'features': [{'id': '0', 'type': 'Feature', 'properties': {'authority':
'Hartlepool', 'names': 0.0, 'registr': 'non-user'}, 'geometry': {'type':
'MultiPolygon', 'coordinates': [(((-1.2684411739451575,
54.726106057183536), (-1.177618431962607, 54.69917558358339),
(-1.1990821541854317, 54.68053931094894), (-1.1608152621574388,
54.63470116747878), (-1.2567026704480637, 54.623338735223086),
(-1.2989994440343908, 54.6279702152551), (-1.3163497415846916,
54.64522360530093), (-1.335046640099353, 54.63131255558137),
(-1.382681311227283, 54.64438440310203), (-1.341239879139743,
54.65020086862752), (-1.3445118278786201, 54.69111059347379),
(-1.2684411739451575, 54.726106057183536)),), (((-1.24214073682385,
54.72239561694049), (-1.2425051040620294, 54.722420353944365),
(-1.2419205226112653, 54.7227132713354), (-1.2419044488925932,
54.722497548992735), (-1.24214073682385, 54.72239561694049)),),
(((-1.2246533829008106, 54.62611617282496), (-1.2246878177957108,
54.62609842464633), (-1.2244912871729203, 54.62617173704485),
(-1.2246533829008106, 54.62611617282496)),)]}, 'bbox': (-1.382681311227283,
54.623338735223086, -1.1608152621574388, 54.726106057183536)}], 'bbox':
(-1.382681311227283, 54.623338735223086, -1.1608152621574388,
54.726106057183536)}, 'type': 'geojson'});

**Comment by jorisvandenbossche** [comment]
@simonm3 thanks for the example! 
So to be clear: this is javascript code, correct? (embedded in html) And how did you include the output of `__geo_interface__` / `_to_geo` in the javascript code? Copying the output in the python console?

**Comment by simonm3** [comment]
I generated the data in python and then combined with an html template using yatl.

Just spotted another difference - there is an id field in the geointerface version but not in the geojson version.

**Comment by simonm3** [comment]
There is a bug in my suggested answer! json.loads returns a dict but we actually want json so need to also call dumps. One difference is the dict has None rather than null.

```
def geojson(x):
    """ return geojson from geodataframe
    includes geometry, crs, properties (from other columns)
    """
    f = BytesIO()
    x.to_file(f, driver="GeoJSON")
    return json.dumps(json.loads(f.getvalue()))
```

**Comment by jorisvandenbossche** [comment]
> I generated the data in python and then combined with an html template using yatl.

I *suppose* then that this is using the str representation of the dictionary, instead of dumping it to json. Because doing `json.dumps(gdf.__geo_interface__())` (or calling `gdf.to_json()`, as that is doing exactly that under the hood) should return valid json (without any trailing comma), while the str repr of a python dict is indeed not guaranteed to be valid json (but this is expected).

**Comment by simonm3** [comment]
Yes it was. However I would imagine the main use case would require geojson
rather than a dict; and one also wants to keep the CRS. I will be using the
fiona version as it is better. IMO main benefit of geopandas is to provide
a single, easy api for a whole bunch of low level applications. As such it
would be good to produce geojson directly rather than having to invoke
fiona directly; especially as it is a bit cryptic to do it in memory with
fiona.

**Comment by simonm3** [comment]
on 2nd thoughts maybe better to just do the loads into a dict as you may want to add other items to a dictionary and then dumps later.

Also the CRS is actually no longer in geojson spec so that is not an issue.

**Comment by jbsetton** [comment]
I needed to iterate through a geodataframe to expose individual geojson objects and noticed the same problems mentioned above (trailing comma and 'id' key in the resulting geojson properties.

I resolved it but my solution is rather 'hacky'.

``` 
gdf_json = json.loads(gdf.to_json())
for f in gdf_json['features']:
    if 'id' in f['properties']:
        del f['properties']['id']  # remove latent 'id' key causing map to fail to render
```
